### ISSUE TITLE
Not able to send to secondary server.

### ISSUE THREAD

**Issue by sachusachin622** [rsyslog/rsyslog#4959]
Hi,

I am trying to do failover scenarios in rsyslog.
I am sending logs to primary server, now in primary server am just doing the packet drop using this command.
**sudo iptables -A INPUT -p tcp --dport 10515 -j DROP**
but after this am expecting rsyslog should send logs to seconadry server but it is not sending logs to secondary server.

Here is the rsyslog.con file content to send logs to remote server.

######################################################
# Ruleset for Log messages to send to Remote server ##
######################################################             
ruleset(name="sendToLogserver" queue.Type="LinkedList" queue.timeoutenqueue="0") {   
#PRIMARY-IP-ADDRESS#                                           
action(type="omfwd" target="1.1.1.1" port="10515"  protocol="tcp" Template="TraditionalFormatWithPRI" )
#SECONDARY-IP-ADDRESS#     
action(type="omfwd" target="2.2.2.2" port="10515"  protocol="tcp" Template="TraditionalFormatWithPRI" action.execOnlyWhenPreviousIsSuspended="on")
}

Please help me to fix this if anything need to do.
Send to secondary server, when primary server is dropping the packets rsyslog should send logs to secondary server.  

Here is the pcap from rsyslog client side seeing TCP retransmission.
here 
rsyslog-client: 10.74.136.52
server: 10.74.136.69

![image](https://user-images.githubusercontent.com/49853320/185047220-4d7ae831-6f4b-4b2a-b396-ffa6eadd4128.png)

After pcap analysis am seeing client is trying to retransmitting the packet and after trying for 15 mins the retransmitation timeout happens on the tcp connection and switches to secondary server. In order to speed this up i tried to apply keepalive.
here is the configuration
######################################################
# Ruleset for Log messages to send to Remote server ##
######################################################
ruleset(name="sendToLogserver" queue.Type="LinkedList" queue.timeoutenqueue="0") {
#PRIMARY-IP-ADDRESS#
action(type="omfwd" target="10.74.136.69" port="10515"  protocol="tcp" keepalive="on" keepalive.probes="3" keepalive.time="30" keepalive.interval="10" Template="TraditionalFormatWithPRI")
#SECONDARY-IP-ADDRESS#
action(type="omfwd" target="10.74.136.96" port="10515"  protocol="tcp" keepalive="on" keepalive.probes="3" keepalive.time="30" keepalive.interval="10" Template="TraditionalFormatWithPRI" action.execOnlyWhenPreviousIsSuspended="on")
} 

with this configuration am expecting the client to send RST packets to primary after 60 seconds and reconnect to secondary server after 60 seconds which is total of 120 seconds after which i should receive logs on the secondary server.
This configuration however does not work and it is falling back to retransmission time out of 15 mins.

Could someone help me to understand configuration is correct if anything missed.
![image](https://user-images.githubusercontent.com/49853320/185046210-51dce7ba-d7e7-4ae8-bb8f-8047e77f17f4.png)

  
Thanks,
Sachin

**Comment by davidelang** [comment]
this will work, but only once the first connection fails and disconnects, which 
will cause a lot of logs to be lost (all the logs that the OS accepts from 
rsyslog before it times out the connection will be considered 'sent' by rsyslog 
as the OS has no way of telling rsyslog "that data I accepted for delivery, I 
can't actually deliver it". Switching to the RELP protocol will eliminate this 
log loss, but it will still take a while to detect that the messages aren't 
getting through and trigger the failover.

watch the logs for the suspend message.

I actually prefer to address this on the receiving side, using something like 
keepalived for simple failover, or pacemake/corosync (same as RHEL high 
availablility package) to manage CLUSTERIP (an iptables feature) along with 
rebindinterval on the senders to do load balancing across multiple machines.

if you want faster failover (and therefor less message loss than TCP), UDP can 
end up more reliable (not as good as RELP, but without the delays that failover 
detection with RELP would introduce)

David Lang

**Comment by rgerhards** [comment]
Apologies for the very late reply. This is probably no longer relevant to the original reporter, but I am adding the conclusion for people who arrive here via search while we work through the backlog. We do not close rsyslog issues lightly, and we are not closing this merely because of age.

Triage result: valid non-code answer.

Reasoning: The thread was answered as TCP failover semantics: the OS may accept data into socket buffers before later discovering the peer is unavailable, so rsyslog cannot reliably fail over every accepted TCP byte. RELP or HA at the receiver side is the recommended pattern.

If this still reproduces on a current rsyslog release and maintained platform, please open a new issue with the current version, complete configuration, logs, and a minimal reproducer so we can investigate it as a fresh current problem.
